codeql/CodeQLParser.g4

```diff
@@ -0,0 +1,720 @@
+/*
+ CodeQL grammar
+ The MIT License (MIT).
+ Copyright (c) 2023, Dijun Liu
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+
+ CodeQL grammar built from the CodeQL Specification https://codeql.github.com/docs/ql-language-reference/ql-language-specification
+*/
+
+parser grammar CodeQLParser;
+
+options { tokenVocab=CodeQLLexer; }
+
+literalId
+    : LOWERID
+    | ATLOWERID
+    | UPPERID
+    | NONE
+    | ANY
+    | keywordAllowed
+    ;
+
+// keywords allowed named in user variable and user predicate decl
+keywordAllowed
+    : 'unique'
+    | 'implements'
+    | 'signature'
+    | 'default'
+    | 'abstract'
+    | 'cached'
+    | 'external'
+    | 'extensible'
+    | 'final'
+    | 'transient'
+    | 'library'
+    | 'private'
+    | 'deprecated'
+    | 'override'
+    | 'additional'
+    | 'query'
+    | 'pragma'
+    | 'inline'
+    | 'inline_late'
+    | 'noinline'
+    | 'nomagic'
+    | 'noopt'
+    | 'assume_small_delta'
+    | 'language'
+    | 'monotonicAggregates'
+    | 'bindingset'
+    | 'only_bind_out'
+    | 'only_bind_into'
+    ;
+
+// keyword can used in predicate call
+keywordPredCallAllowed
+    : keywordAllowed
+    | 'any'
+    | 'none'
+    ;
+
+// varname ::= lowerId
+varname
+    : LOWERID
+    | keywordAllowed
+    ;
+
+//ql ::= QL_DOC? moduleBody
+ql
+    : QL_DOC? moduleBody
```

Comment:
Yes sir. I'll add EOF at the end of the root rule

Reply:
The build fails for PHP because of a "symbol conflict". In Antlr, some targets have problems with certain symbol names that other targets don't have a problem with. "CLASS" in the lexer .g4 and "class_" in the parser .g4 are two of those with PHP. These symbols could be renamed to "CLASS_" and "class_" respectively, but after trying that, I see the target "works" for a few inputs quickly (e.g., examples/annotation.ql), but is ridiculously slow for the entire test suite in examples/. The build has a 5 minute time limit for any step, so it will still fail.

Preferably, please remove the ["PHP;" from the desc.xml](https://github.com/flyinox/grammars-v4/blob/16adde8bea8edf3454f014125df4c41bcd343ebb/codeql/desc.xml#L3C48-L3C52).